src/number_theory/pell.lean

```diff
@@ -398,4 +427,252 @@ end solution₁
+of any positive solution. -/
+lemma le_x {a₁ : solution₁ d} (h : is_fundamental a₁) {a : solution₁ d} (hax : 1 < a.x) :
+  a₁.x ≤ a.x :=
+h.2.2 hax
+
+/-- The `y`-coordinate of a fundamental solution is a lower bound for the `y`-coordinate
+of any positive solution. -/
+lemma le_y {a₁ : solution₁ d} (h : is_fundamental a₁) {a : solution₁ d} (hax : 1 < a.x)
+  (hay : 0 < a.y) :
+  a₁.y ≤ a.y :=
+begin
+  have H : d * (a₁.y ^ 2 - a.y ^ 2) = a₁.x ^ 2 - a.x ^ 2 := by { rw [a.prop_x, a₁.prop_x], ring },
+  rw [← abs_of_pos hay, ← abs_of_pos h.2.1, ← sq_le_sq, ← mul_le_mul_left h.d_pos, ← sub_nonpos,
+      ← mul_sub, H, sub_nonpos, sq_le_sq, abs_of_pos (zero_lt_one.trans h.1),
+      abs_of_pos (zero_lt_one.trans hax)],
+  exact h.le_x hax,
+end
+
+/-- If we multiply a positive solution with the inverse of a fundamental solution,
+the `x`-coordinate decreases and the `y`-coordinate remains nonnegative. -/
+lemma mul_inv_lt_x {a₁ : solution₁ d} (h : is_fundamental a₁) {a : solution₁ d} (hax : 1 < a.x)
+  (hay : 0 < a.y) :
+  0 < (a * a₁⁻¹).x ∧ (a * a₁⁻¹).x < a.x ∧ 0 ≤ (a * a₁⁻¹).y :=
+begin
+  have hax' := zero_lt_one.trans hax,
+  have H₁ : a.x * a₁.y ≤ a.y * a₁.x,
+  { rw [← abs_of_pos hax', ← abs_of_pos hay,
+        ← abs_of_pos h.x_pos, ← abs_of_pos h.2.1, ← abs_mul, ← abs_mul, ← sq_le_sq,
+        mul_pow, mul_pow, a.prop_x, a₁.prop_x, ← sub_nonneg],
+    ring_nf,
+    rw [sub_nonneg, sq_le_sq, abs_of_pos hay, abs_of_pos h.2.1],
+    exact h.le_y hax hay, },
+  simp only [x_mul, x_inv, y_inv, mul_neg, add_neg_lt_iff_le_add',
+             lt_add_neg_iff_add_lt, zero_add, y_mul, le_neg_add_iff_add_le, add_zero],
+  refine ⟨_, _, H₁⟩,
+  { refine (mul_lt_mul_left hax').mp _,
+    rw [(by ring : a.x * (d * (a.y * a₁.y)) = (d * a.y) * (a.x * a₁.y))],
+    refine ((mul_le_mul_left $ mul_pos h.d_pos hay).mpr H₁).trans_lt _,
+    rw [← mul_assoc, mul_assoc d, ← sq, a.prop_y, ← sub_pos],
+    ring_nf,
+    exact zero_lt_one.trans h.1, },
+  { refine (mul_lt_mul_left h.2.1).mp _,
+    rw [(by ring : a₁.y * (a.x * a₁.x) = a.x * a₁.y * a₁.x)],
+    refine ((mul_le_mul_right $ zero_lt_one.trans h.1).mpr H₁).trans_lt _,
+    rw [mul_assoc, ← sq, a₁.prop_x, ← sub_neg],
+    ring_nf,
+    rw [sub_neg, ← abs_of_pos hay, ← abs_of_pos h.2.1, ← abs_of_pos hax', ← abs_mul, ← sq_lt_sq,
+        mul_pow, a.prop_x],
+    calc
+      a.y ^ 2 = 1 * a.y ^ 2                  : (one_mul _).symm
+          ... ≤ d * a.y ^ 2                  : (mul_le_mul_right $ sq_pos_of_pos hay).mpr h.d_pos
+          ... < d * a.y ^ 2 + 1              : lt_add_one _
+          ... = (1 + d * a.y ^ 2) * 1        : by rw [add_comm, mul_one]
+          ... ≤ (1 + d * a.y ^ 2) * a₁.y ^ 2
+                : (mul_le_mul_left (by {have := h.d_pos, positivity})).mpr (sq_pos_of_pos h.2.1), }
+end
+
+/-- Any nonnegative solution is a power with nonnegative exponent of a fundamental solution. -/
+lemma pow_of_nonneg {a₁ : solution₁ d} (h : is_fundamental a₁) {a : solution₁ d} (hax : 0 < a.x)
```

Comment:
```suggestion
lemma eq_pow_of_nonneg {a₁ : solution₁ d} (h : is_fundamental a₁) {a : solution₁ d} (hax : 0 < a.x)
```

Reply:
OK.

Reply:
I'm also changing `zpow_or_neg_zpow` to `eq_zpow_or_neg_zpow`.